File: bindings/python/src/libsonare/_types_engine.py
"""Public type definitions for libsonare.

camelCase property aliases mirror the JS binding's public API so users moving
between languages see the same names. They intentionally violate PEP8 N802.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

from ._types_analysis import (
    AutomationCurve,
    EngineTelemetryError,
    EngineTelemetryType,
    SectionType,
    TimeSignature,
)


@dataclass(frozen=True, slots=True)
class MixMeterSnapshot:
    """Realtime mixer meter snapshot for one strip."""

    peak_db_l: float
    peak_db_r: float
    rms_db_l: float
    rms_db_r: float
    correlation: float
    mono_compat_width: float
    mono_compat_peak: float
    mono_compat_side_rms: float
    likely_mono_compatible: bool
    momentary_lufs: float
    short_term_lufs: float
    integrated_lufs: float
    gain_reduction_db: float
    true_peak_db_l: float
    true_peak_db_r: float
    max_true_peak_db: float
    seq: int
    # Per-plane surround meters (5.1/7.1). Each tuple holds channel_count values;
    # indices 0/1 mirror the *_l/*_r stereo fields above.
    channel_count: int = 0
    peak_db: tuple[float, ...] = ()
    rms_db: tuple[float, ...] = ()
    true_peak_db: tuple[float, ...] = ()


@dataclass(frozen=True, slots=True)
class GoniometerPoint:
    """A single left/right sample pair for goniometer (vectorscope) display."""

    left: float
    right: float


@dataclass(frozen=True, slots=True)
class MixResult:
    """Result of rendering a small stereo mixer scene."""

    left: list[float]
    right: list[float]
    sample_rate: int
    meters: list[MixMeterSnapshot]


@dataclass(frozen=True, slots=True)
class EngineTelemetry:
    """Realtime engine telemetry event."""

    type: EngineTelemetryType
    error: EngineTelemetryError
    render_frame: int
    timeline_sample: int
    audible_timeline_sample: int
    graph_latency_samples_q8: int
    value: int

    @property
    def renderFrame(self) -> int:  # noqa: N802
        return self.render_frame

    @property
    def timelineSample(self) -> int:  # noqa: N802
        return self.timeline_sample

    @property
    def audibleTimelineSample(self) -> int:  # noqa: N802
        return self.audible_timeline_sample

    @property
    def graphLatencySamplesQ8(self) -> int:  # noqa: N802
        return self.graph_latency_samples_q8


class EngineTrackMonitorMode(IntEnum):
    """Per-track-lane monitor tap mode for the realtime engine."""

    OFF = 0
    PFL = 1
    AFL = 2


@dataclass(frozen=True, slots=True)
class ParameterInfo:
    """DAW parameter metadata for automation/introspection UIs."""

    id: int
    name: str
    unit: str
    min_value: float
    max_value: float
    default_value: float
    rt_safe: bool
    default_curve: AutomationCurve


@dataclass(frozen=True, slots=True)
class AutomationPoint:
    """PPQ automation breakpoint."""

    ppq: float
    value: float
    curve_to_next: AutomationCurve = AutomationCurve.LINEAR


class MarkerKind(IntEnum):
    """Timeline marker kind. Mirrors SonareMarkerKind in the C ABI and the
    other bindings' marker-kind enums; the values are part of the ABI."""

    MARKER = 0
    TEXT = 1
    LYRIC = 2
    CUE_POINT = 3
    KEY_SIGNATURE = 4


@dataclass(frozen=True, slots=True)
class EngineMarker:
    """Timeline marker used by the realtime engine transport.

    ``kind`` is a :class:`MarkerKind` ordinal; ``key_fifths`` (-7..7, sharps
    positive) and ``key_minor`` apply only to the key-signature kind.
    """

    id: int
    ppq: float
    name: str = ""
    kind: int = 0
    key_fifths: int = 0
    key_minor: bool = False


@dataclass(frozen=True, slots=True)
class ProjectMarker:
    """Timeline marker stored on a headless :class:`~libsonare._project.Project`.

    Same shape as :class:`EngineMarker`: ``kind`` is a :class:`MarkerKind`
    ordinal and ``key_fifths`` / ``key_minor`` apply only to the key-signature
    kind.
    """

    id: int
    ppq: float
    name: str = ""
    kind: int = 0
    key_fifths: int = 0
    key_minor: bool = False


@dataclass(frozen=True, slots=True)
class ProjectTrack:
    """Read-only, stored-order project track descriptor."""

    id: int
    kind: int
    midi_destination_id: int
    gain: float
    pan: float
    mute: bool
    solo: bool
    name: str


@dataclass(frozen=True, slots=True)
class ProjectClip:
    """Read-only, stored-order project clip descriptor."""

    id: int
    track_id: int
    source_id: int
    source_kind: int
    start_ppq: float
    length_ppq: float
    source_offset_ppq: float
    gain: float
    loop_mode: int
    loop_length_ppq: float


@dataclass(frozen=True, slots=True)
class ProjectSource:
    """Read-only, stored-order project source descriptor."""

    id: int
    kind: int
    channel_count: int
    storage_handle_id: int
    sample_rate_hint: float
    name_or_uri: str
    content_hash: str = ""
    external_stem_role: str = ""


@dataclass(frozen=True, slots=True)
class EngineMetronomeConfig:
    """Realtime engine metronome click configuration."""

    enabled: bool = False
    beat_gain: float = 0.35
    accent_gain: float = 0.7
    # Explicit click length in samples (0..384000). Zero selects the
    # sample-rate-derived duration below.
    click_samples: int = 0
    # Click duration in seconds (0..1); used when click_samples is 0. 0.0 selects
    # the engine default (2 ms).
    click_seconds: float = 0.0


@dataclass(frozen=True, slots=True)
class EngineClip:
    """Owned audio clip schedule for realtime engine playback."""

    id: int
    channels: list[list[float]] | None
    start_ppq: float
    track_id: int = 0
    length_samples: int | None = None
    clip_offset_samples: int = 0
    loop: bool = False
    gain: float = 1.0
    fade_in_samples: int = 0
    fade_out_samples: int = 0
    warp_mode: str | int = "off"
    warp_anchors: list[tuple[float, float]] | None = None
    page_provider: object | None = None


@dataclass(frozen=True, slots=True)
class EngineMidiEvent:
    """Absolute render-frame MIDI event for realtime engine MIDI clips."""

    render_frame: int
    word0: int = 0
    word1: int = 0
    word2: int = 0
    word3: int = 0
    word_count: int = 0
    group: int = 0
    sysex_handle: int = 0


@dataclass(frozen=True, slots=True)
class EngineMidiClipSchedule:
    """Compiled realtime MIDI clip schedule."""

    events: list[EngineMidiEvent]
    id: int = 0
    track_id: int = 0
    # MIDI destination to route to. None (the default) falls back to track_id;
    # an explicit value, including 0, is forwarded unchanged.
    destination_id: int | None = None
    start_sample: int = 0
    start_ppq: float = 0.0
    length_samples: int = 0
    loop: bool = False
    loop_length_samples: int = 0


@dataclass(frozen=True, slots=True)
class ClipPageRequest:
    """Paged clip sample request drained from the realtime engine."""

    clip_id: int
    channel: int
    sample: int


@dataclass(frozen=True, slots=True)
class EngineCaptureStatus:
    """Capture progress for the realtime engine recording sink."""

    captured_frames: int
    overflow_count: int
    armed: bool
    punch_enabled: bool
    source: str
    record_offset_samples: int


@dataclass(frozen=True, slots=True)
class EngineBounceOptions:
    """Offline export options for the realtime engine."""

    total_frames: int
    block_size: int = 128
    num_channels: int = 2
    target_sample_rate: int = 48000
    source_sample_rate: int = 48000
    normalize_lufs: bool = False
    target_lufs: float = -14.0
    dither: int = 0
    dither_bits: int = 16
    dither_seed: int = 0


@dataclass(frozen=True, slots=True)
class EngineBounceResult:
    """Interleaved offline export result from the realtime engine."""

    interleaved: list[float]
    frames: int
    num_channels: int
    sample_rate: int
    integrated_lufs: float


@dataclass(frozen=True, slots=True)
class EngineFreezeOptions:
    """Offline freeze options for replacing current engine output with a clip."""

    total_frames: int
    block_size: int = 128
    num_channels: int = 2
    clip_id: int = 1
    start_ppq: float = 0.0
    gain: float = 1.0


@dataclass(frozen=True, slots=True)
class EngineFreezeResult:
    """Result of freezing current engine output into a scheduled clip."""

    clip_id: int
    frames: int
    num_channels: int


class EngineGraphNodeType(IntEnum):
    """Builtin processor node type for realtime engine graphs."""

    PASS_THROUGH = 0
    GAIN = 1


class EngineGraphMix(IntEnum):
    """Connection mix mode for realtime engine graphs."""

    REPLACE = 0
    ADD = 1


@dataclass(frozen=True, slots=True)
class EngineGraphNode:
    """Prepared realtime engine graph node."""

    id: str
    type: EngineGraphNodeType = EngineGraphNodeType.PASS_THROUGH
    gain_db: float = 0.0
    num_ports: int = 0


@dataclass(frozen=True, slots=True)
class EngineGraphConnection:
    """Prepared realtime engine graph connection."""

    source_node: str
    source_port: int
    dest_node: str
    dest_port: int
    mix: EngineGraphMix = EngineGraphMix.ADD


@dataclass(frozen=True, slots=True)
class EngineGraphParameterBinding:
    """Map an engine automation parameter id to a graph node processor."""

    param_id: int
    node_id: str


@dataclass(frozen=True, slots=True)
class EngineGraphSpec:
    """Prepared realtime engine graph specification."""

    nodes: list[EngineGraphNode]
    connections: list[EngineGraphConnection]
    input_node: str
    output_node: str
    num_channels: int = 2
    parameter_bindings: list[EngineGraphParameterBinding] | None = None


@dataclass(frozen=True, slots=True)
class MeterTelemetryRecord:
    """A meter snapshot drained from the realtime engine's meter tap."""

    target_id: int
    render_frame: int
    seq: int
    peak_db_l: float
    peak_db_r: float
    rms_db_l: float
    rms_db_r: float
    true_peak_db_l: float
    true_peak_db_r: float
    max_true_peak_db: float
    correlation: float
    mono_compat_width: float
    momentary_lufs: float
    short_term_lufs: float
    integrated_lufs: float
    gain_reduction_db: float
    dropped_records: int


@dataclass(frozen=True, slots=True)
class MeterTelemetryRecordWide:
    """A per-plane meter snapshot for a surround mix target.

    ``peak_db``/``rms_db``/``true_peak_db`` carry ``channel_count`` valid planes
    in canonical WAVE order (5.1 = L R C LFE Ls Rs, 7.1 = L R C LFE Ls Rs Lss
    Rss). Use this drain for a surround target; ``MeterTelemetryRecord`` stays the
    stereo fast path.
    """

    target_id: int
    render_frame: int
    seq: int
    channel_count: int
    peak_db: list[float]
    rms_db: list[float]
    true_peak_db: list[float]
    max_true_peak_db: float
    correlation: float
    mono_compat_width: float
    momentary_lufs: float
    short_term_lufs: float
    integrated_lufs: float
    gain_reduction_db: float
    dropped_records: int


@dataclass(frozen=True, slots=True)
class ScopeTelemetryRecord:
    """A spectrum/vectorscope snapshot drained from the realtime engine.

    Each vectorscope point is an ``(left, right)`` tuple — the idiomatic Python
    shape. The Node / WASM surfaces expose the same data as ``{left, right}``
    objects; this representation difference is intentional, the values match.
    """

    target_id: int
    render_frame: int
    seq: int
    dropped_records: int
    bands: list[float]
    points: list[tuple[float, float]]


@dataclass(frozen=True, slots=True)
class ExternalMidiEvent:
    """One lowered MIDI 1.0 message drained from the external-MIDI output queue.

    Each event is a single MIDI 1.0 byte message (``bytes`` is 1..3 bytes). A
    ``destination_id`` of ``0xFFFFFFFF`` (4294967295) tags a clock/transport byte
    forwarded for external tempo sync; any other value is the MIDI destination id
    the event was routed to.
    """

    destination_id: int
    render_frame: int
    bytes: bytes


@dataclass(frozen=True, slots=True)
class TransportState:
    """Read-only snapshot of the realtime engine transport state."""

    playing: bool
    looping: bool
    render_frame: int
    sample_position: int
    ppq_position: float
    bpm: float
    loop_start_ppq: float
    loop_end_ppq: float
    sample_rate: float
    # Musical position derived from the tempo map (computed every block).
    bar_start_ppq: float
    bar_count: int
    time_signature: TimeSignature
    # One-based beat within the current bar; ``bar_count`` above is zero-based.
    beat: int
    # Fractional position within the current beat, in [0, 1).
    beat_fraction: float


@dataclass(frozen=True, slots=True)
class Section:
    """A detected song-structure section."""

    type: SectionType
    start: float
    end: float
    energy_level: float
    confidence: float
    canonical_name: str = ""

    @property
    def name(self) -> str:
        if self.canonical_name:
            return self.canonical_name
        names = {
            SectionType.INTRO: "Intro",
            SectionType.VERSE: "Verse",
            SectionType.PRE_CHORUS: "Pre-Chorus",
            SectionType.CHORUS: "Chorus",
            SectionType.BRIDGE: "Bridge",
            SectionType.INSTRUMENTAL: "Instrumental",
            SectionType.OUTRO: "Outro",
            SectionType.UNKNOWN: "Unknown",
        }
        return names[self.type]


@dataclass(frozen=True, slots=True)
class SectionResult:
    """Song-structure analysis result."""

    sections: list[Section]


@dataclass(frozen=True, slots=True)
class MelodyPoint:
    """A single point on a melody contour."""

    time: float
    frequency: float
    confidence: float


@dataclass(frozen=True, slots=True)
class MelodyResult:
    """Melody contour analysis result."""

    points: list[MelodyPoint]
    pitch_range_octaves: float
    pitch_stability: float
    mean_frequency: float
    vibrato_rate: float

File: bindings/node/src/types_engine.ts
import type { TimeSignature } from './types_analysis.js';
import type { SendTiming } from './types_mastering.js';
import type { ProjectWarpAnchor, WarpMode } from './types_project.js';

export type EngineTelemetryType = 0 | 1;

export type EngineTelemetryError =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15
  | 16;

export interface EngineTelemetry {
  type: EngineTelemetryType;
  error: EngineTelemetryError;
  renderFrame: number;
  timelineSample: number;
  audibleTimelineSample: number;
  graphLatencySamplesQ8: number;
  value: number;
}

/** Meter telemetry record drained from {@link RealtimeEngine.drainMeterTelemetry}. */
export interface EngineMeterTelemetry {
  /** Meter tap target id (e.g. master/bus identifier). */
  targetId: number;
  /** Render-frame timestamp of the snapshot. */
  renderFrame: number;
  /** Monotonic sequence number. */
  seq: number;
  /** Left-channel peak level in dB. */
  peakDbL: number;
  /** Right-channel peak level in dB. */
  peakDbR: number;
  /** Left-channel RMS level in dB. */
  rmsDbL: number;
  /** Right-channel RMS level in dB. */
  rmsDbR: number;
  /** Left-channel true-peak level in dB. */
  truePeakDbL: number;
  /** Right-channel true-peak level in dB. */
  truePeakDbR: number;
  /** Maximum true-peak across channels in dB. */
  maxTruePeakDb: number;
  /** Stereo correlation in `[-1, 1]`. */
  correlation: number;
  /** Mono-compatibility width metric. */
  monoCompatWidth: number;
  /** Momentary loudness (LUFS). */
  momentaryLufs: number;
  /** Short-term loudness (LUFS). */
  shortTermLufs: number;
  /** Integrated loudness (LUFS). */
  integratedLufs: number;
  /** Gain reduction in dB. */
  gainReductionDb: number;
  /** Number of records dropped before this snapshot. */
  droppedRecords: number;
}

/**
 * One lowered MIDI 1.0 message drained from
 * {@link RealtimeEngine.drainExternalMidi}. A single queued channel-voice event
 * may lower to more than one message (e.g. a MIDI 2.0 program change with bank
 * select), so the drain returns one entry per lowered message.
 */
export interface EngineExternalMidiEvent {
  /**
   * Originating track MIDI destination id, or the transport sentinel
   * `0xFFFFFFFF` (4294967295) for clock / start / continue / stop bytes meant
   * for every external port.
   */
  destinationId: number;
  /**
   * Render-frame coordinate of the event. Channel-voice events carry the
   * timeline sample position; clock/transport bytes carry the device render
   * frame.
   */
  renderFrame: number;
  /** MIDI 1.0 status + data bytes (1..3 entries). */
  bytes: number[];
}

/**
 * Per-plane meter telemetry record drained from
 * {@link RealtimeEngine.drainMeterTelemetryWide} for a surround target. The
 * `peakDb`/`rmsDb`/`truePeakDb` arrays carry `channelCount` planes in canonical
 * WAVE order (5.1 = L R C LFE Ls Rs, 7.1 = L R C LFE Lss Rss Ls Rs).
 */
export interface EngineMeterTelemetryWide {
  /** Meter tap target id (e.g. master/bus identifier). */
  targetId: number;
  /** Render-frame timestamp of the snapshot. */
  renderFrame: number;
  /** Monotonic sequence number. */
  seq: number;
  /** Number of valid per-plane meters (1..8). */
  channelCount: number;
  /** Per-plane peak level in dB (length `channelCount`). */
  peakDb: number[];
  /** Per-plane RMS level in dB (length `channelCount`). */
  rmsDb: number[];
  /** Per-plane true-peak level in dB (length `channelCount`). */
  truePeakDb: number[];
  /** Maximum true-peak across channels in dB. */
  maxTruePeakDb: number;
  /** Stereo correlation in `[-1, 1]` (front pair). */
  correlation: number;
  /** Mono-compatibility width metric (front pair). */
  monoCompatWidth: number;
  /** Momentary loudness (LUFS). */
  momentaryLufs: number;
  /** Short-term loudness (LUFS). */
  shortTermLufs: number;
  /** Integrated loudness (LUFS). */
  integratedLufs: number;
  /** Gain reduction in dB. */
  gainReductionDb: number;
  /** Number of records dropped before this snapshot. */
  droppedRecords: number;
}

/** Scope telemetry record drained from {@link RealtimeEngine.drainScopeTelemetry}. */
export interface EngineScopeTelemetry {
  /** Scope tap target id (0 = master, `laneIndex + 1` for lanes, `33 + busIndex` for buses). */
  targetId: number;
  /** Render-frame timestamp of the snapshot. */
  renderFrame: number;
  /** Monotonic sequence number. */
  seq: number;
  /** Number of records dropped before this snapshot. */
  droppedRecords: number;
  /** FFT magnitude spectrum in dBFS, linear-spaced over `[0, Nyquist]`. */
  bands: number[];
  /**
   * Goniometer/vectorscope sample points as `{ left, right }` objects. The
   * Python surface exposes the same data as `(left, right)` tuples; this
   * representation difference is intentional, the values match.
   */
  points: { left: number; right: number }[];
}

/** Read-only engine transport snapshot from {@link RealtimeEngine.getTransportState}. */
export interface EngineTransportState {
  /** Whether the transport is currently playing. */
  playing: boolean;
  /** @deprecated Use `playing`; kept for one release for compatibility. */
  isPlaying?: boolean;
  /** Whether looping is enabled. */
  looping: boolean;
  /** Current render-frame counter. */
  renderFrame: number;
  /** Current timeline position in samples. */
  samplePosition: number;
  /** Current position in pulses-per-quarter-note. */
  ppq: number;
  /** Current tempo in beats per minute. */
  bpm: number;
  /** Loop start in PPQ. */
  loopStartPpq: number;
  /** Loop end in PPQ. */
  loopEndPpq: number;
  /** Engine sample rate in Hz. */
  sampleRate: number;
  /** PPQ of the current bar's downbeat (derived from the tempo map). */
  barStartPpq: number;
  /** Zero-based index of the current bar. */
  barCount: number;
  /** Time signature in effect at the current PPQ. */
  timeSignature: TimeSignature;
  /** One-based beat within the current bar (`barCount` is zero-based). */
  beat: number;
  /** Fractional position within the current beat, in [0, 1). */
  beatFraction: number;
}

/**
 * Engine automation breakpoint curve as an integer code.
 * Canonical ordinals (matches mixer `AutomationCurve`):
 *   0 = Linear (default), 1 = Exponential, 2 = Hold, 3 = SCurve.
 */
export type EngineAutomationPointCurve =
  | 0
  | 1
  | 2
  | 3
  | 'linear'
  | 'exponential'
  | 'hold'
  | 's-curve';

export interface EngineParameterInfo {
  id: number;
  name: string;
  unit: string;
  minValue: number;
  maxValue: number;
  defaultValue: number;
  rtSafe: boolean;
  defaultCurve: EngineAutomationPointCurve;
}

export interface EngineAutomationPoint {
  ppq: number;
  value: number;
  curveToNext?: EngineAutomationPointCurve;
}

/** Structured marker kind ordinals; mirror SonareMarkerKind / SmfMarkerKind. */
export enum MarkerKind {
  Marker = 0,
  Text = 1,
  Lyric = 2,
  CuePoint = 3,
  KeySignature = 4,
}

export interface EngineMarker {
  id: number;
  ppq: number;
  name?: string;
  /** Marker kind (MarkerKind ordinal); defaults to MarkerKind.Marker. */
  kind?: number;
  /** Key signature only: -7..7 (sharps positive). */
  keyFifths?: number;
  /** Key signature only: false = major, true = minor. */
  keyMinor?: boolean;
}

export interface ProjectMarker {
  id: number;
  ppq: number;
  name: string;
  /** Marker kind (MarkerKind ordinal). */
  kind: number;
  /** Key signature only: -7..7 (sharps positive). */
  keyFifths: number;
  /** Key signature only: false = major, true = minor. */
  keyMinor: boolean;
}

export interface EngineMetronomeConfig {
  enabled: boolean;
  beatGain?: number;
  accentGain?: number;
  clickSamples?: number;
  /** Click duration in seconds; used when clickSamples is 0 to derive the click length from the sample rate. */
  clickSeconds?: number;
}

export interface EngineClip {
  id: number;
  trackId?: number;
  channels?: Float32Array[];
  startPpq: number;
  lengthSamples?: number;
  clipOffsetSamples?: number;
  loop?: boolean;
  gain?: number;
  fadeInSamples?: number;
  fadeOutSamples?: number;
  warpMode?: WarpMode;
  warpAnchors?: ProjectWarpAnchor[];
  pageProvider?: number | { readonly id: number };
}

/**
 * Speaker bed layout for a bus or source, mirroring the C enum
 * `SonareChannelLayout`: `0` = mono, `1` = stereo, `2` = 5.1 (L R C LFE Ls Rs),
 * `3` = 7.1 (L R C LFE Lss Rss Ls Rs).
 */
export type ChannelLayout = 0 | 1 | 2 | 3;

export interface EngineTrackLane {
  trackId: number;
  sends?: EngineTrackSend[];
  /**
   * Bus the lane's post-fader output sums into instead of the master mix
   * (group/folder routing); 0 or absent keeps the lane on the master mix.
   */
  outputBusId?: number;
  /**
   * Input channel layout of the source feeding this lane. Absent defaults to
   * stereo. Stored but inert until the surround DSP path lands.
   */
  sourceChannelLayout?: ChannelLayout;
}

export interface EngineTrackSend {
  busId: number;
  levelDb?: number;
  enabled?: boolean;
  /**
   * Tap point for the send relative to the lane fader. Absent defaults to
   * post-fader, preserving the behavior before this field existed.
   */
  sendTiming?: SendTiming | number;
}

export interface EngineBus {
  busId: number;
  gainDb?: number;
  /**
   * Channel layout of this bus. Absent defaults to stereo. The master bus
   * carries the project output layout. Stored but inert until the surround DSP
   * path lands.
   */
  channelLayout?: ChannelLayout;
}

export interface ClipPageRequest {
  clipId: number;
  channel: number;
  sample: number;
}

export interface FileClipPageProviderOptions {
  numChannels: number;
  numSamples: number;
  pageFrames: number;
  dataOffsetBytes?: number;
}

export interface EngineCaptureStatus {
  capturedFrames: number;
  overflowCount: number;
  armed: boolean;
  punchEnabled: boolean;
  source: 'output' | 'input';
  recordOffsetSamples: number;
}

export type EngineCaptureSource = EngineCaptureStatus['source'] | number;

export interface EngineBounceOptions {
  totalFrames: number;
  blockSize?: number;
  numChannels?: number;
  targetSampleRate?: number;
  sourceSampleRate?: number;
  normalizeLufs?: boolean;
  targetLufs?: number;
  dither?: 0 | 1 | 2 | 3;
  ditherBits?: number;
  ditherSeed?: number;
}

export interface EngineBounceResult {
  interleaved: Float32Array;
  frames: number;
  numChannels: number;
  sampleRate: number;
  integratedLufs: number;
}

export interface EngineFreezeOptions {
  totalFrames: number;
  blockSize?: number;
  numChannels?: number;
  clipId?: number;
  startPpq?: number;
  gain?: number;
}

export interface EngineFreezeResult {
  clipId: number;
  frames: number;
  numChannels: number;
}

export type EngineGraphNodeType = 0 | 1;

/**
 * Mixing intent for a graph edge (`0` = replace, `1` = add).
 *
 * NOTE: not currently honored — the compiled graph always sums edges into a
 * shared destination port in an order-independent way (the first edge into a
 * port overwrites, every later edge adds), regardless of this value. Retained
 * for API compatibility and to express intent; multiple edges into one port are
 * always summed.
 */
export type EngineGraphMix = 0 | 1;

export interface EngineGraphNode {
  id: string;
  type?: EngineGraphNodeType;
  gainDb?: number;
  numPorts?: number;
}

export interface EngineGraphConnection {
  sourceNode: string;
  sourcePort: number;
  destNode: string;
  destPort: number;
  mix?: EngineGraphMix;
}

export interface EngineGraphParameterBinding {
  paramId: number;
  nodeId: string;
}

export interface EngineGraphSpec {
  nodes: EngineGraphNode[];
  connections: EngineGraphConnection[];
  inputNode: string;
  outputNode: string;
  numChannels?: number;
  parameterBindings?: EngineGraphParameterBinding[];
}

/** Options for {@link RealtimeEngine.bindMidiCc}. All fields are optional. */
export interface MidiCcBindOptions {
  /** Lower end of the mapped parameter range. Default `0`. */
  minValue?: number;
  /** Upper end of the mapped parameter range. Default `1`. */
  maxValue?: number;
}

/** One absolute render-frame MIDI event accepted by {@link RealtimeEngine.setMidiClips}. */
export interface EngineMidiEvent {
  renderFrame: number;
  /** First UMP word. `data0` is accepted by native wrappers as an alias. */
  word0?: number;
  /** Second UMP word. `data1` is accepted by native wrappers as an alias. */
  word1?: number;
  word2?: number;
  word3?: number;
  wordCount?: number;
  /** UMP group, 0..15. */
  group?: number;
  /** Optional SysEx side-store handle for native hosts. */
  sysexHandle?: number;
  /** Project-style alias for one-word MIDI 1.0 events. */
  data0?: number;
  /** Project-style alias for the second UMP word. */
  data1?: number;
}

/** One compiled realtime MIDI clip schedule accepted by {@link RealtimeEngine.setMidiClips}. */
export interface EngineMidiClipSchedule {
  id?: number;
  trackId?: number;
  destinationId?: number;
  startSample?: number;
  startPpq?: number;
  lengthSamples?: number;
  loop?: boolean;
  loopLengthSamples?: number;
  events: EngineMidiEvent[];
}
